oak_idl/src/lib.rs

```diff
@@ -0,0 +1,185 @@
+struct Handle(u32);
+
+impl Transport for Handle {
+    fn invoke(&self, request_message: &Message) -> Result<Message, TransportError> {
+        // The serialization of the outer message is fixed by the transport, since the header needs
+        // to be deserializable by the other end of the transport, but the body of the
+        // request / response is serialized at a higher level, and in potentially different format,
+        // since that's handled by the application.
+        let request_bytes =
+            bincode::serialize(&request_message).map_err(|_| TransportError::InvalidMessage)?;
+        let response_bytes = invoke(self.0, &request_bytes)?;
+        let response_message =
+            bincode::deserialize(&response_bytes).map_err(|_| TransportError::InvalidMessage)?;
+        Ok(response_message)
+    }
+}
+
+/// A low-level message-oriented transport over channels. This is meant to emulate the low-level
+/// Wasm ABI in Oak, and will eventually be replaced by that.
+fn invoke(_channel_handle: u32, _request: &[u8]) -> Result<Vec<u8>, TransportError> {
+    unimplemented!()
+}
+
+/// This macro generates the following objects (assuming it is invoked with the first argument set
+/// to `Name`):
+///
+/// - a struct named `NameClient`, exposing a method for each method defined in the macro.
+/// - a struct named `NameServer`, which implements the [`Transport`] trait, dispatching each
+///   request to the appropriate method on the underlying service implementation.
+/// - a trait named `Name`, with a method for each method defined in the macro, and an additional
+///   default method named `serve` which returns an instance of `NameServer`; the developer of a
+///   service would usually define a concrete struct and manually implement this trait for it.
+#[macro_export]
+macro_rules! service {
+    (
+        $name:ident {
+            $( $method_id:literal => fn $method:ident ( $request_arg:ident : $request_type:ty ) -> $response_type:ty ; )*
+        }
+    ) => {
+        // We use <https://docs.rs/paste/latest/paste/> in order to generate identifiers with
+        // different suffixes.
+        paste::paste! {
+            pub struct [<$name "Client">]<T: $crate::Transport> {
+                transport: T,
+            }
+
+            impl <T: $crate::Transport>[<$name "Client">]<T> {
+                pub fn new(transport: T) -> Self {
+                    Self {
+                        transport
+                    }
+                }
+
+                $(
+                    pub fn $method(&self, $request_arg: $request_type) -> Result<$response_type, $crate::ClientError> {
+                        let request_body = bincode::serialize(&$request_arg).map_err(|_| $crate::ClientError::InvalidRequest)?;
+                        let request_message = $crate::Message {
+                            header: $crate::Header {
+                                transaction_id: 0,
```

Comment:
I assume setting this to 0 is temporary. Will this be incremented automatically by the client?